feat(content-sidebar): added support for active comment reply

kkuliczkowski-box opened the pull request:
Added support for handling links with active item that is a reply. This is to support mentioning users directly in a reply. The link that user will receive in notification email will have only type and id of the feed item. This change covers that scenario by doing additional request for that replie's details (in order to get id of its parent). Logic can be seen in the diagram:

![handling-active-comment-reply](https://user-images.githubusercontent.com/105496865/200943320-c07e66fb-6153-4e65-9266-2a088efd308b.png)

**Please note that this is a temporary solution done based on time constraints presented by Product. Final solution would be to change the link in notification email to also include `parentId`, thus making this additional request mentioned above not needed.**

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -110,23 +125,38 @@ class ActivityFeed extends React.Component<Props, State> {
-        const { isInputOpen: currIsInputOpen } = this.state;
+        const { feedItems: currFeedItems, activeFeedEntryId, onShowReplies = noop } = this.props;
+        const { activeEntry, isInputOpen: prevIsInputOpen } = prevState;
+        const { activeEntry: prevActiveEntry, isInputOpen: currIsInputOpen } = this.state;
```
Is this inverted? It looks like this should be `currActiveEntry` and the line above should be `prevActiveEntry`.

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -110,23 +125,38 @@ class ActivityFeed extends React.Component<Props, State> {
-        const { feedItems: currFeedItems, activeFeedEntryId } = this.props;
-        const { isInputOpen: prevIsInputOpen } = prevState;
-        const { isInputOpen: currIsInputOpen } = this.state;
+        const { feedItems: currFeedItems, activeFeedEntryId, onShowReplies = noop } = this.props;
```
Let's use defaultProps for the default `noop` value here.

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -110,23 +125,38 @@ class ActivityFeed extends React.Component<Props, State> {
             this.resetFeedScroll();
         }
 
-        if (didLoadFeedItems || hasActiveFeedEntryIdChanged) {
```
Do we still need to check `didLoadFeedItems`? If not, won't this fail in most circumstances?

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -88,12 +99,16 @@ type Props = {
 
 class ActivityFeed extends React.Component<Props, State> {
     state = {
+        activeEntry: undefined,
```
We should avoid storing derived state from props in this way, since it can lead to synchronization issues. Let's retrieve the entry as needed using the value from `props.activeFeedEntryId`.

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
         return feedItems.length === 0 || (feedItems.length === 1 && feedItems[0].type === ItemTypes.fileVersion);
     };
 
+    getActiveEntry = (callback: (activeEntry: FeedItem | null, repliesNeedToBeLoaded?: boolean) => void) => {
```
Can we use a promise or async/await instead of introducing callbacks?

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
+        );
+        // If the active entry is a first level Feed item, return it
+        if (firstLevelItem) {
+            return callback(firstLevelItem);
```
We generally want to avoid multiple returns, as it makes a function very difficult to debug and reason about. This function is also _very_ complicated and probably needs to be decomposed.

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
+                !item.replies ||
+                !Array.isArray(item.replies)
+            ) {
+                return false;
```
Should this return undefined?

kkuliczkowski-box replied:
I think It can return `false` as this is the return statement of the `find` callback function. I've copied that piece of logic into `ActivitySidebar`.

jstoffan commented on src/api/ThreadedComments.js:
```diff
@@ -228,6 +228,45 @@ class ThreadedComments extends MarkerBasedApi {
+        permissions: BoxItemPermission,
+        successCallback: (comment: Comment) => void,
+    }): void {
+        this.errorCode = ERROR_CODE_FETCH_COMMENTS;
```
Should this use a new error code?

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
+            return callback(firstLevelItemWithActiveReply);
+        }
+
+        if (!getComment) {
```
Is this possible with the current implementation? It looks like the method is always defined.

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -43,6 +49,11 @@ type Props = {
     file: BoxItem,
     getApproverWithQuery?: Function,
     getAvatarUrl: GetAvatarUrlCallback,
+    getComment?: (
```
Should we handle this case in `ActivitySidebar.fetchFeedItems` rather than inside ActivityFeed?

kkuliczkowski-box replied:
That makes a lot of sense. I've moved the logic for fetching reply data to `ActivitySidebar` but I've placed it in `fetchFeedItemsSuccessCallback` as I didn't want to change the original purpose of `fetchFeedItems. It seemed better to put it in `fetchFeedItemsSuccessCallback` as I'm taking actions based on the success of the `fetchFeedItems`.

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -298,7 +378,7 @@ class ActivityFeed extends React.Component<Props, State> {
                     className="bcs-activity-feed-items-container"
                 >
-                    {isLoading && (
+                    {(isLoading || isReplyDataLoading) && (
```
The combination of loading state here indicates again that this should probably occur when the feed items are being fetched rather than separately. If not, though, should this move into the `hasLoaded` method?

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -313,8 +393,8 @@ class ActivityFeed extends React.Component<Props, State> {
                             {...activityFeedError}
-                            activeFeedEntryId={activeFeedEntryId}
-                            activeFeedEntryType={activeFeedEntryType}
+                            activeFeedEntryId={activeEntry?.id}
```
Let's revert these changes given that we want to avoid derived state.

tjuanitas replied:
is there a reason we call the state `activeEntry` instead of `activeFeedEntry`?

tjuanitas commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
+
+        this.setState({ isReplyDataLoading: true });
+        return getComment(
+            activeFeedEntryId,
+            ({ parent }) => {
+                this.setState({ isReplyDataLoading: false });
+                const parentItem = feedItems.find(({ id }) => id === parent?.id) || null;
+                callback(parentItem, true);
+            },
+            () => {
+                this.setState({ isReplyDataLoading: false });
+                callback(null);
+            },
```
where is `getComment` initially from passed from? i'm curious how the order of these args are determined.

kkuliczkowski-box commented:
I've changed the approach and followed @jstoffan's advice and moved the whole logic of fetching the reply data to `ActivitySidebar`. Thanks to that we don't need this additional "loading" step and the reply data is being loaded before the feed items are displayed. Because all of the changes done in `ActivityFeed` have been reverted, I haven't responded to most of the comments.

I'm struggling with Flow check at the moment and will continue working on it.

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -152,20 +158,31 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
             endMarkName: MARK_NAME_JS_READY,
         });
-        this.state = {};
+        this.state = { activeFeedEntryId, activeFeedEntryType, approverSelectorContacts: [] };
```
We should almost always avoid duplicating our props or any prop-derived value in state, as it can easily lead to synchronization issues. Let's update the logic to use the values directly from props.

kkuliczkowski-box replied:
I understand your point but in this case I don't see any alternative really. The `activeFeedEntryId` and `activeFeedEntryType` props represent the values of url params. Previously they were simply passed to `ActivityFeed`, where logic for displaying "selected state" lives.

Now with the possibility of a reply id being set as `activeFeedEntryId` we need to check if it is somewhere within feed items. If it turns out that it's a reply we need to pass its parent as `activeFeedEntryId` as the parent needs to be "selected/highlighted". In that case I'm updating values in state and pass them to `ActivityFeed`. If no action is taken. Moreover if the reply id is not found within feed items, we need to fetch for that reply's data in order to find its parent and then set its id in state. 

Ultimately this is a temporary solution until we get the `partentId` as an url param. With that we would not need to store the active item id in state as it would simply be either the `itemId` or the `parentId`.

kkuliczkowski-box commented on src/api/Feed.js:
```diff
@@ -462,6 +462,51 @@ class Feed extends Base {
+     * @param {Comment} comment - comment data
+     * @return {void}
+     */
+    fetchThreadedCommentSuccessCallback = (resolve: Function, successCallback: Function, comment: Comment): void => {
```
I needed to keep `comment` as the last param because I'm binding this method in `fetchThreadedComment` method.

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -717,7 +738,34 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+        this.getActiveFeedEntryData(feedItems)
+            .then(({ id, type }) => {
+                this.setState({
+                    ...baseChanges,
```
I'm a bit concerned about race conditions here. What happens if `fetchFeedItems` is called again while this promise is still in progress? Would we wind up with stale feed items? It seems like we'd want to handle this conditional promise chain prior to the success callback being invoked?

kkuliczkowski-box replied:
These additional asynchronous actions (fetching reply data and its parent replies) will take place only once when feed items are not yet loaded. From what I can tell, during that time fetchFeedItems will not be called as the whole Activity Sidebar is in the "loading state".

I've also moved that additional actions logic outside of `fetchFeedItemsSuccessCallback`.

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -717,7 +738,34 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
-        this.setState({ feedItems, activityFeedError: undefined });
+        const baseChanges = { feedItems, activityFeedError: undefined };
+
+        // If activeFeedEntryId is not set or
```
This comment is largely describing the logic below. We should use comments to explain "why" something is written if it's not obvious in the code.

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -827,6 +875,149 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+                    return;
+                }
+                resolve({ id: activeFeedEntryId, type: activeFeedEntryType });
+                return;
```
We should avoid multiple returns, as it makes logic very difficult to debug and reason about. We're doing quite a bit of duplicate work in this function and it would be great to refactor it for simplicity.

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -827,6 +875,149 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+     * @param {ErrorCallback} onError -the error callback
+     * @return {void}
+     */
+    getComment = (
```
Minor, but is this needed? Can we call `fetchThreadedComment` directly, instead?

jstoffan commented on src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:
```diff
@@ -127,6 +140,20 @@ class ActivityFeed extends React.Component<Props, State> {
+            currFeedItems &&
+            !isItemInFeed(currFeedItems, activeFeedEntryReplyId)
+        ) {
+            onShowReplies(activeFeedEntryId, activeFeedEntryType);
```
This seems very indirect. Can we load the replies if we detect that they're needed within the `fetchFeedItems` chain?

kkuliczkowski-box replied:
Yes, that was a wrong approach. I've reverted the `ActivityFeed` completely and fetching for replies is now a part of the `fetchFeedItems` chain during the first load of `ActivitySidebar`.

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -530,11 +536,32 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+        if (
+            activeFeedEntryId &&
+            replies.length === 1 &&
+            feedItems.find(
```
Should this be `.some`? It looks we do a form of this check in quite a few places. Would it be meaningful to move it to a utility function?

kkuliczkowski-box replied:
Right, I've changed to `some` and also noticed that I could actually use a method (`getCommentFeedItemByReplyId`) I've created earlier, that already has logic to find the parent based on reply id and have used it now.
I've merged this PR because of time constraints but opened a new one for these minor fixes: https://github.com/box/box-ui-elements/pull/3148

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -530,11 +536,32 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+        // that is in currently being updated parent, in order to disable active item
+        if (
+            activeFeedEntryId &&
+            replies.length === 1 &&
```
Would this cause any issues with threads that are exactly one reply long?

kkuliczkowski-box replied:
It should not as threads can be "hidden" only if there are 2 or more replies. Hiding replies basically means setting replies array to just one entry (the last reply).

jstoffan commented on src/elements/content-sidebar/ActivitySidebar.js:
```diff
@@ -671,26 +698,68 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+                const parentType: CommentFeedItemType =
+                    type === FEED_ITEM_TYPE_COMMENT ? FEED_ITEM_TYPE_COMMENT : FEED_ITEM_TYPE_ANNOTATION;
+
+                this.getFeedItemsWithReplies(feedItems, id, parentType)
```
Can we flatten out this promise to avoid needing the duplicate `.catch`?

```
this.getActiveFeedEntryData(feedItems)
    .then(({ id, type }) => {
        if (
            !id ||
            !type ||
            this.isActiveEntryInFeed(feedItems, activeFeedEntryId) ||
            !this.isItemTypeComment(type)
        ) {
            return Promise.resolve(feedItems);
        }

        const parentType: CommentFeedItemType =
            type === FEED_ITEM_TYPE_COMMENT ? FEED_ITEM_TYPE_COMMENT : FEED_ITEM_TYPE_ANNOTATION;

        return this.getFeedItemsWithReplies(feedItems, id, parentType)
    })
    .then(updatedItems => this.fetchFeedItemsSuccessCallback(updatedItems))
    .catch(error => this.fetchFeedItemsErrorCallback(feedItems, [error]));

jstoffan reviewed (approved):
LGTM as a temporary solution. We'll want to revisit this code in the near future, though.

kkuliczkowski-box commented:
I've created a new PR in order to address the latest comments: https://github.com/box/box-ui-elements/pull/3148
